presto-hive/src/test/java/com/facebook/presto/hive/TestOrcPageSourceMemoryTracking.java

```diff
@@ -0,0 +1,524 @@
+
+        assertEquals(driverContext.getSystemMemoryUsage(), 0);
+
+        long memoryUsage = -1;
+        for (int i = 0; i < 20; i++) {
+            assertFalse(operator.isFinished());
+            Page page = operator.getOutput();
+            assertNotNull(page);
+            page.getBlock(1);
+            if (memoryUsage == -1) {
+                memoryUsage = driverContext.getSystemMemoryUsage();
+            }
+            else {
+                assertEquals(driverContext.getSystemMemoryUsage(), memoryUsage);
+            }
+        }
+        assertBetweenInclusive(memoryUsage, 460000L, 469999L);
+
+        memoryUsage = -1;
+        for (int i = 20; i < 40; i++) {
+            assertFalse(operator.isFinished());
+            Page page = operator.getOutput();
+            assertNotNull(page);
+            page.getBlock(1);
+            if (memoryUsage == -1) {
+                memoryUsage = driverContext.getSystemMemoryUsage();
+            }
+            else {
+                assertEquals(driverContext.getSystemMemoryUsage(), memoryUsage);
+            }
+        }
+        assertBetweenInclusive(memoryUsage, 460000L, 469999L);
+
+        memoryUsage = -1;
+        for (int i = 40; i < 50; i++) {
+            assertFalse(operator.isFinished());
+            Page page = operator.getOutput();
+            assertNotNull(page);
+            page.getBlock(1);
+            if (memoryUsage == -1) {
+                memoryUsage = driverContext.getSystemMemoryUsage();
+            }
+            else {
+                assertEquals(driverContext.getSystemMemoryUsage(), memoryUsage);
+            }
+        }
+        assertBetweenInclusive(memoryUsage, 360000L, 369999L);
+
+        Assert.assertFalse(operator.isFinished());
+        assertNull(operator.getOutput());
+        Assert.assertTrue(operator.isFinished());
+        assertEquals(driverContext.getSystemMemoryUsage(), 0);
+    }
+
+    @Test
+    public void testScanFilterAndProjectOperator()
+            throws Exception
+    {
+        DriverContext driverContext = testPreparer.newDriverContext();
+        SourceOperator operator = testPreparer.newScanFilterAndProjectOperator(driverContext);
+
+        assertEquals(driverContext.getSystemMemoryUsage(), 0);
+
+        for (int i = 0; i < 52; i++) {
+            assertFalse(operator.isFinished());
+            operator.getOutput();
+            assertBetweenInclusive(driverContext.getSystemMemoryUsage(), 600000L, 639999L);
+        }
+
+        for (int i = 52; i < 65; i++) {
+            assertFalse(operator.isFinished());
+            operator.getOutput();
+            assertBetweenInclusive(driverContext.getSystemMemoryUsage(), 500000L, 539999L);
+        }
+
+        // Page source is over, but data still exist in buffer of ScanFilterProjectOperator
+        Assert.assertFalse(operator.isFinished());
+        assertNull(operator.getOutput());
+        assertBetweenInclusive(driverContext.getSystemMemoryUsage(), 170000L, 179999L);
```

Comment:
All these constants that are in here. Are they guaranteed by the file format, or might they change if we improve our reader in the future? If the latter, there should probably be a comment saying it's safe to change them if they break in the future